File: backend/models/questionModel.js
const mongoose = require('mongoose');

const questionSchema = new mongoose.Schema({
  order: {
    type: Number,
    default: 0,
  },
  questionText: {
    type: String,
    required: [true, 'A question must have text'],
  },
  questionType: {
    type: String,
    enum: ['mcq', 'true-false', 'ai-practice'],
    default: 'mcq',
  },
  marks: {
    type: Number,
    required: [true, 'A question must have marks'],
    default: 1,
  },
  options: {
    type: [
      {
        text: {
          type: String,
          required: [true, 'An option must have text'],
        },
        isCorrect: {
          type: Boolean,
          required: [true, 'An option must specify if it is correct'],
        },
      },
    ],
    required: function () {
      return this.questionType !== 'ai-practice';
    },
    validate: {
      validator(val) {
        if (!val || val.length === 0) return true;
        return val.length >= 2;
      },
      message: 'A question must have at least 2 options',
    },
  },
  expectedSign: {
    type: String,
    required: function () {
      return this.questionType === 'ai-practice';
    },
  },
  expectedType: {
    type: String,
    enum: ['letter', 'number', 'word', 'sentence'],
    default: 'letter',
  },
  imageUrl: {
    type: String,
  },
  quizId: {
    type: mongoose.Schema.ObjectId,
    ref: 'Quiz',
    required: [true, 'A question must belong to a quiz'],
  },
}, { timestamps: true });

questionSchema.index({ quizId: 1, order: 1, createdAt: 1 });

const Question = mongoose.model('Question', questionSchema);

module.exports = Question;

File: mobile/lib/models/question_model.dart
class QuestionOptionModel {
  final String id;
  final String text;
  final bool isCorrect;

  const QuestionOptionModel({
    required this.id,
    required this.text,
    required this.isCorrect,
  });

  factory QuestionOptionModel.fromJson(Map<String, dynamic> json) {
    return QuestionOptionModel(
      id: json['_id'] ?? json['id'] ?? '',
      text: json['text'] ?? '',
      isCorrect: json['isCorrect'] as bool? ?? false,
    );
  }
}

class QuestionModel {
  final String id;
  final String questionText;
  final String questionType; // 'mcq' | 'true-false' | 'ai-practice'
  final int marks;
  final List<QuestionOptionModel> options;
  final String? imageUrl;
  final String? expectedSign;
  final String? quizId;

  const QuestionModel({
    required this.id,
    required this.questionText,
    required this.questionType,
    required this.marks,
    required this.options,
    this.imageUrl,
    this.expectedSign,
    this.quizId,
  });

  factory QuestionModel.fromJson(Map<String, dynamic> json) {
    final rawOptions = json['options'] as List? ?? [];
    return QuestionModel(
      id: json['_id'] ?? json['id'] ?? '',
      questionText: json['questionText'] ?? '',
      questionType: json['questionType'] ?? 'mcq',
      marks: json['marks'] as int? ?? 1,
      options: rawOptions
          .map((e) => QuestionOptionModel.fromJson(e as Map<String, dynamic>))
          .toList(),
      imageUrl: json['imageUrl'],
      expectedSign: json['expectedSign'],
      quizId: json['quizId'] is String ? json['quizId'] : null,
    );
  }
}
